internal/extender/resource.go

```diff
@@ -336,92 +337,66 @@ func (s *SparkSchedulerExtender) potentialNodes(availableNodesSchedulingMetadata
-				}
-				return node, outcome, err
-			}
-			softReservation := v1beta1.Reservation{
-				Node:   node,
-				CPU:    sparkResources.executorResources.CPU,
-				Memory: sparkResources.executorResources.Memory,
-			}
-			err = s.softReservationStore.AddReservationForPod(ctx, driver.Labels[common.SparkAppIDLabel], executor.Name, softReservation)
+	}
+
+	unboundReservationNodes, err := s.resourceReservationManager.FindUnboundReservationNodes(ctx, executor)
+	if err != nil {
+		return "", failureInternal, werror.WrapWithContextParams(ctx, err, "error when looking for unbound reservations")
+	}
+	if len(unboundReservationNodes) > 0 {
+		if resultNode, ok := s.getReservationNodeFromNodeList(unboundReservationNodes, nodeNames); ok {
+			err := s.resourceReservationManager.ReserveForExecutor(ctx, executor, resultNode)
 			if err != nil {
-				return "", failureInternal, err
+				return "", failureInternal, werror.WrapWithContextParams(ctx, err, "failed to reserve node for executor")
 			}
-			// We might have created a demand object for this executor when we were under min count, so we should remove if it exists
 			s.removeDemandIfExists(ctx, executor)
-			return node, successScheduledExtraExecutor, nil
+			return resultNode, success, nil
 		}
-		return "", outcome, unboundResErr
-	}
-	// the reservation to be selected for the current executor needs to be on a node that the extender has received from kube-scheduler
-	nodeToReservation := make(map[string]string, len(unboundReservations))
-	for _, reservationName := range unboundReservations {
-		nodeToReservation[resourceReservation.Spec.Reservations[reservationName].Node] = reservationName
 	}
-	var unboundReservation string
-	for _, name := range nodeNames {
-		if reservation, ok := nodeToReservation[name]; ok {
-			unboundReservation = reservation
-			break
-		}
+
+	// Else, check if you still can have an executor, and if yes, reschedule
+	freeExecutorSpots, err := s.resourceReservationManager.GetFreeExecutorSpots(ctx, executor)
+	if err != nil {
+		return "", failureInternal, werror.WrapWithContextParams(ctx, err, "error when checking for free executor spots")
 	}
-	copyResourceReservation := resourceReservation.DeepCopy()
-
-	if unboundReservation == "" {
-		// no nodes for the unbound reservations exists in nodeNames, this might be because the node for the reservations are terminated
-		// try to reschedule the executor, breaking FIFO, but preventing executor starvation
-		// we are guaranteed len(unboundResourceReservations) > 0
-		unboundReservation = unboundReservations[0]
-		driver, err := s.podLister.getDriverPod(ctx, executor)
-		if err != nil {
-			return "", failureInternal, err
-		}
-		sparkResources, err := sparkResources(ctx, driver)
+	if freeExecutorSpots > 0 {
+		isExtraExecutor := len(unboundReservationNodes) > 0 // Dynamic allocation executor since there were no free unbound reservations
```

Comment:
what's a case where freeExecutorSpots is greater than zero, it's not dynamic allocation, and there are no unbound reservations?

Reply:
Good point. The initial intention was to cover the case where there are unbound reservations, but they were not passed to us as potential nodes, so we would get in a situation where we get here with `freeExecutorSpots > 0` even though there are unbound reservations previously found.
I notice though that we can't consider this a dynamic allocation executor, so will fix this.

Reply:
Fixed to check the opposite, i.e. that there are no unbound reservations found